mlflow/utils/model_utils.py

```diff
@@ -428,3 +429,69 @@ def _validate_pyfunc_model_config(model_config):
             "valid file path or of type ``dict`` with string keys.",
             error_code=INVALID_PARAMETER_VALUE,
         )
+
+
+RECORD_ENV_VARS = {
+    # api key related
+    "API_KEY",
+    # databricks auth related
+    "DATABRICKS_HOST",
+    "DATABRICKS_USERNAME",
+    "DATABRICKS_PASSWORD",
+    "DATABRICKS_TOKEN",
+    "DATABRICKS_INSECURE",
+    "DATABRICKS_CLIENT_ID",
+    "DATABRICKS_CLIENT_SECRET",
+    "_DATABRICKS_WORKSPACE_HOST",
+    "_DATABRICKS_WORKSPACE_ID",
+    # dbconnect related
+    "SPARK_REMOTE",
+    "SPARK_LOCAL_REMOTE",
+}
+
+
+class EnvTracker(dict):
+    """
+    Track environment variables set and accessed during the context manager's lifetime.
+    Note that if the environment variable does not exist, it will not be tracked.
+    """
+
+    def __init__(self, *args, **kwargs):
+        self.env_vars = set()
+        super().__init__(*args, **kwargs)
+
+    def __setitem__(self, key, value):
+        super().__setitem__(key, value)
+        if any(env_var in key for env_var in RECORD_ENV_VARS):
+            self.env_vars.add(key)
+
+    def __getitem__(self, key):
+        if key in self and any(env_var in key for env_var in RECORD_ENV_VARS):
+            self.env_vars.add(key)
+        return super().__getitem__(key)
+
+    def get(self, key, *args, **kwargs):
+        if key in self and any(env_var in key for env_var in RECORD_ENV_VARS):
+            self.env_vars.add(key)
+        return super().get(key, *args, **kwargs)
+
+    def get_env_vars(self):
+        return self.env_vars
+
+
+@contextlib.contextmanager
+def env_var_tracker():
+    """
+    Context manager for temporarily tracking environment variables accessed.
+    It tracks both environment variables set and accessed during the context manager's lifetime.
+    """
+    from mlflow.environment_variables import MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING
+
+    if MLFLOW_RECORD_ENV_VARS_IN_MODEL_LOGGING.get():
+        try:
+            os.environ = EnvTracker(os.environ)
+            yield os.environ
+        finally:
+            os.environ = {**os.environ}
```

Comment:
Is this really safe? `os.environ` is not a object, it's like a special mapping object.

Reply:
Yea good point, I think we can mock os.environ's get/set method to collect them in a set directly. Let me update

Reply:
Updated to patch os._Environ's methods